File: DesignPatterns/01_GOF23/DP_examples/Go/01_Creational/Builder/Director.go
package Builder

type Director[Builder IMealBuilder] struct {
	mealBuilder *Builder
}

func (d *Director[Builder]) Construct() MealBuilder {
	if d.mealBuilder == nil {
		d.mealBuilder = new(Builder)
	}
	return MealBuilder{
		builders: (*d.mealBuilder).GetBuilders(),
	}
}

File: DesignPatterns/01_GOF23/DP_examples/Lua/01_Creational/Builder/Director.lua
-- Director.lua
-- 指挥者

local Builder = require("ConcreteBuilder")
local Product = require("Product")

--- 构造套餐
---@param mealBuilder IMealBuilder
---@return MealBuilder
local function Construct(mealBuilder)
    local builders = mealBuilder:GetBuilders()
    return Builder.NewMealBuilder(builders)
end
return {
    Construct = Construct,
    VegMealBuilder = Builder.VegMealBuilder,
    BurgerBuilder = Builder.BurgerBuilder,  
    ItemKind = Product.ItemKind
}
